internal/pkg/agent/application/coordinator/state/state.go

```diff
@@ -246,79 +266,141 @@ func (cs *CoordinatorState) State() (s State) {
+	//
+	// The ordering we use is:
+	// - first, the listener on subscribeChan
+	// - second, the listener on stateChangedChan
+	// - after that, two cases for each subscriber, in the same order as the
+	//   subscribers array: first its done channel, then its listener channel.
+	//
+	// All subscribers are included in the array of select cases even when some
+	// have already been updated, that way we don't need to worry about the
+	// order changing. Instead, subscribers that have already been updated have
+	// the listener channel of their select case set to nil.
+	selectCases := []reflect.SelectCase{
+		{
+			Dir:  reflect.SelectRecv,
+			Chan: reflect.ValueOf(cs.subscribeChan),
+		},
+		{
+			Dir:  reflect.SelectRecv,
+			Chan: reflect.ValueOf(cs.stateChangedChan),
+		},
+	}
+	const newSubscriberIndex = 0
+	const stateChangedIndex = 1
+	const firstSubscriberIndex = 2
+
+	currentState := cs.State()
+
+	// resetListeners is called when a state change notification arrives, to
+	// reactivate the listener channels of all subscribers and update their
+	// select case to the new state value.
+	resetListeners := func() {
+		currentState = cs.State()
+		for i, subscriber := range subscribers {
+			listenerIndex := firstSubscriberIndex + 2*i + 1
+			selectCases[listenerIndex].Chan = reflect.ValueOf(subscriber.ch)
+			selectCases[listenerIndex].Send = reflect.ValueOf(currentState)
+		}
+	}
+
+	// addSubscriber is a helper to add a new subscriber and its select
+	// cases to our lists.
+	addSubscriber := func(subscriber StateSubscription) {
+		subscribers = append(subscribers, subscriber)
+		selectCases = append(selectCases,
+			// Add a select case receiving from the subscriber's done channel
+			reflect.SelectCase{
+				Dir:  reflect.SelectRecv,
+				Chan: reflect.ValueOf(subscriber.ctx.Done()),
+			},
+			// Add a select case sending to the subscriber's listener channel
+			reflect.SelectCase{
+				Dir:  reflect.SelectSend,
+				Chan: reflect.ValueOf(subscriber.ch),
+				Send: reflect.ValueOf(currentState),
+			})
+	}
 
-	send := func(sub *StateSubscription) {
-		t := time.NewTimer(time.Second)
-		defer t.Stop()
+	for {
+		// Always try a standalone receive on the state changed channel first, so
+		// it gets priority over other updates.
 		select {
-		case <-sub.ctx.Done():
-		case sub.ch <- state:
-		case <-t.C:
-			// subscriber didn't read from the channel after 1 second; so we unblock
+		case <-cs.stateChangedChan:
+			resetListeners()
+		default:
 		}
-	}
 
-	for _, sub := range cs.subscribe {
-		send(sub)
+		chosen, value, _ := reflect.Select(selectCases)
+		if chosen == stateChangedIndex {
+			resetListeners()
+		} else if chosen == newSubscriberIndex {
+			subscriber, ok := value.Interface().(StateSubscription)
+			if ok {
+				addSubscriber(subscriber)
+			}
+		} else {
+			subscriberIndex := (chosen - firstSubscriberIndex) / 2
+			if (chosen-firstSubscriberIndex)%2 == 0 {
+				// The subscriber's done channel has been closed, remove
```

Comment:
I don't think we have a test where a subscriber is removed.

I went to measure the test coverage of this code manually and the Go tooling won't give me a test report for this package because the state implementation is alone in its own package. We should just move this into the coordinator package or move the tests into the state package to make this possible.

Reply:
It was moved into its own package to ensure that the coordinator only interfaces with the state machine using the public interface. This is to ensure that locks and other pieces of the state are not touched directly by the coordinator.